Line/android/LINELoginDemo/app/src/main/java/com/google/firebase/linelogindemo/util/LineLoginHelper.java

```diff
@@ -0,0 +1,255 @@
+                        break;
+                    case CANCELED: // Login canceled by user
+                        Exception e = new Exception("User cancelled LINE Login.");
+                        source.setException(e);
+                        break;
+                    default: /* Error */ {
+                        Throwable cause = future.getCause();
+                        if (cause instanceof LineSdkLoginException) {
+                            LineSdkLoginException loginException = (LineSdkLoginException) cause;
+                            Log.e(TAG, loginException.getMessage());
+                            source.setException(loginException);
+                        } else {
+                            source.setException(new Exception("Unknown error occurred in LINE SDK."));
+                        }
+                    } break;
+                }
+            }
+        });
+
+        return source.getTask();
+    }
+
+    private Task<String> getFirebaseAuthToken(Context context, final String lineAccessToken) {
+        final TaskCompletionSource<String> source = new TaskCompletionSource<>();
+
+        // STEP 2: Exchange LINE access token for Firebase Custom Auth token
+        HashMap<String, String> validationObject = new HashMap<>();
+        validationObject.put("token", lineAccessToken);
+
+        // Exchange LINE Access Token for Firebase Auth Token
+        Response.Listener<JSONObject> responseListener = new Response.Listener<JSONObject>() {
+            @Override
+            public void onResponse(JSONObject response) {
+                try {
+                    String firebaseToken = response.getString("firebase_token");
+                    Log.d(TAG, "Firebase Token = " + firebaseToken);
+                    source.setResult(firebaseToken);
+                } catch (Exception e) {
+                    source.setException(e);
+                }
+            }
+        };
+
+        Response.ErrorListener errorListener = new Response.ErrorListener() {
+            @Override
+            public void onErrorResponse(VolleyError error) {
+                Log.e(TAG, error.toString());
+                source.setException(error);
+            }
+        };
+
+        JsonObjectRequest fbTokenRequest = new JsonObjectRequest(
+                Request.Method.POST, mLineAcessscodeVerificationEndpoint,
+                new JSONObject(validationObject),
+                responseListener, errorListener);
+
+        NetworkSingleton.getInstance(context).addToRequestQueue(fbTokenRequest);
+
+        return source.getTask();
+    }
+
+    private Task<FirebaseUser> loginWithFirebase(Activity activity, String firebaseToken) {
+        final TaskCompletionSource<FirebaseUser> source = new TaskCompletionSource<>();
+
+        // STEP 3: Use Firebase Custom Auth token to login Firebase
+        FirebaseAuth auth = FirebaseAuth.getInstance();
+        auth.signInWithCustomToken(firebaseToken)
```

Comment:
Why not just have this method return the `Task<AuthResult>` directly rather than wrapping it in another task?